### PR TITLE
feat(app-cmds): support Range/String for min/max value/length options

### LINKED ISSUE #590: `Range` for slash option ranges

### PR DESCRIPTION
**Pull request by ooliver1**
## Summary

Closes #590
Adds support for a shorthand syntax for range validation.

### Supports

- `Range[1, 2]` for min and max ranges.
- `Range[1.1, 2.2]` for floating-point ranges.
- `Range[2]` for max ranges only.
- `Range[1, ...]` for min ranges only.

#### Supported Syntax

```py
@bot.slash_command()
async def range_int(inter: Interaction, option: Range[1, 2]):
    await inter.send(f"You inputted {option}")


@bot.slash_command()
async def range_float(inter: Interaction, option: Range[1.1, 2.2]):
    await inter.send(f"You inputted {option}")


@bot.slash_command()
async def only_max(inter: Interaction, option: Range[2]):
    await inter.send(f"You inputted {option}")


@bot.slash_command()
async def only_min(inter: Interaction, option: Range[3, ...]):
    await inter.send(f"You inputted {option}")

@bot.slash_command()
async def annotated_range(inter: Interaction, option: Annotated[float, Range[1.1, 2.2]]):
    await inter.send(f"You inputted {option}")

@bot.slash_command()
async def string_length(inter: Interaction, option: String[1, 2]):
    await inter.send(f"You inputted {option}")
```

integer range
float range
max range
min range
annotated range
string length

##### Not Setting Either ``min`` or ``max``

```py
@bot.slash_command()
async def neither_defined(inter: Interaction, option: Range[...]):
    ...
```

```py
TypeError: At least one of min or max must be set.
```

##### Not Using ``float`` or ``int`` items

```py
@bot.slash_command()
async def invalid_items(inter: Interaction, option: Range["1", "2"]):
    ...
```

```py
TypeError: Range min must be int or float.
```

##### Using More Than 2 Items

```py
@bot.slash_command()
async def too_many_items(inter: Interaction, option: Range[1, 2, 3]):
    ...
```

```py
ValueError: Range can only take 1-2 arguments
```

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by EmreTech — APPROVED** [review]
Haven't tested, but the code looks good

File: nextcord/application_command.py [hunk 1]
```diff
@@ -1317,6 +1325,8 @@ def __init__(
         annotation_choices: List[Union[str, int, float]] = []
         annotation_channel_types: List[ChannelType] = []
         annotation_converters: List[OptionConverter] = []
+        annotation_min_value: Optional[float] = MISSING
+        annotation_max_value: Optional[float] = MISSING
```
Comment (DenverCoder1) [inline nextcord/application_command.py:1329]:
```suggestion
        annotation_min_value: Optional[Union[float, int]] = MISSING
        annotation_max_value: Optional[Union[float, int]] = MISSING
```

File: nextcord/application_command.py [hunk 2]
```diff
@@ -1430,8 +1460,16 @@ def __init__(
...
-        self.max_value = cmd_arg.max_value
+        if annotation_min_value is not MISSING:
+            self.min_value = annotation_min_value
+        else:
+            self.min_value = cmd_arg.min_value
+
+        if annotation_max_value is not MISSING:
+            self.max_value = annotation_max_value
+        else:
+            self.max_value = cmd_arg.max_value
```
Comment (DenverCoder1) [inline nextcord/application_command.py:1471]: `cmd_arg` should take precedence here for consistency.

If it's explicitly set in the SlashOption, the annotation should be disregarded.

```suggestion
        self.min_value = annotation_min_value if annotation_min_value is not MISSING else None
        if cmd_arg.min_value is not None:
            self.min_value = cmd_arg.min_value
        
        self.max_value = annotation_max_value if annotation_max_value is not MISSING else None
        if cmd_arg.max_value is not None:
            self.max_value = cmd_arg.max_value
```

**Review by DenverCoder1 — COMMENTED** [review]
1. Maybe `Range[..., max]` should also be supported for consistency with using `...` as a max?
2. Maybe something similar for string length range should be added at the same time?
3. I tried using this with `Annotated` and got [an error](https://paste.nextcord.dev/?id=1660171449888173)

**Comment by ooliver1** [comment]
> 1. Maybe `Range[..., max]` should also be supported for consistency with using `...` as a max?
> 2. Maybe something similar for string length range should be added at the same time?
> 3. I tried using this with `Annotated` and got [an error](https://paste.nextcord.dev/?id=1660171449888173)

1. 7f9ac0b067d601d1a49776b94a9c8725a61e8492
2. e75b60d21afa9be1bdf45eeea71ba7d5643e3ab8
3. 999af8270acda8070cdfa0cee9926f59b815e8fa

File: nextcord/application_command.py [hunk 3]
```diff
@@ -3276,3 +3293,177 @@ def unpack_annotation(
...
+        else:
+            raise TypeError("Range max must be int or float.")
+
+        if Inner.min is None and Inner.max is None:
+            raise TypeError("At least one of min or max must be set.")
+
+        return Inner
+
+
+class String(int):
```
Comment (alentoghostflame) [inline nextcord/application_command.py:3398]: It isn't THAT important, but why does it subclass `int`?

**Review by alentoghostflame — APPROVED** [review]
Haven't tested, but mostly looks good to me.

File: nextcord/application_command.py (hunk 3 above)
Comment (ooliver1) [reply nextcord/application_command.py:3398]: I suppose it doesnt need to anymore as `__class_getitem__` returns that, legacy code

Comment (DenverCoder1) [reply nextcord/application_command.py:3398]: Would it help to subclass `str` for typing purposes or not really?

Comment (ooliver1) [reply nextcord/application_command.py:3398]: `__class_getitem__` pretends to return `str/int/float` so the types are recognised like that. This was done in the early stage of this PR when this used a metaclass

File: nextcord/application_command.py [hunk 4]
```diff
@@ -3276,3 +3293,177 @@ def unpack_annotation(
...
+        value: Union[
+            int,
+            Tuple[int, int],
+            Tuple[int, EllipsisType],
+            Tuple[EllipsisType, int],
+        ],
+    ) -> Type[int]:
+        class Inner(String, OptionConverter):
+            def __init__(self):
+                super().__init__(option_type=str)
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3432]: Currently a NotImplemented error is raised when running a command with Range

```suggestion
        class Inner(String, OptionConverter):
            def __init__(self):
                super().__init__(option_type=str)
                
            async def convert(self, interaction: Interaction, value: Any) -> Any:
                return value
```

File: nextcord/application_command.py [hunk 5]
```diff
@@ -3276,3 +3293,177 @@ def unpack_annotation(
...
+            float,
+            Tuple[int, int],
+            Tuple[float, float],
+            Tuple[Union[int, float], EllipsisType],
+            Tuple[EllipsisType, Union[int, float]],
+        ],
+    ) -> Type[Union[int, float]]:
+        class Inner(Range, OptionConverter):
+            def __init__(self):
+                super().__init__(option_type=type(self.min or self.max))
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3358]:
```suggestion
        class Inner(Range, OptionConverter):
            def __init__(self):
                super().__init__(option_type=type(self.min or self.max))
                
            async def convert(self, interaction: Interaction, value: Any) -> Any:
                return value
```

File: nextcord/application_command.py (hunk 3 above)
Comment (ooliver1) [inline nextcord/application_command.py:3398]:
```suggestion
class String:
```

File: nextcord/application_command.py [hunk 6]
```diff
@@ -3276,3 +3293,177 @@ def unpack_annotation(
         raise ValueError(f"Given Annotation {given_annotation} has an unhandled origin: {origin}")
 
     return type_ret, literal_ret
+
+
+class Range(int):
```
Comment (ooliver1) [inline nextcord/application_command.py:3298]:
```suggestion
class Range:
```

File: nextcord/application_command.py [hunk 7]
```diff
@@ -3276,3 +3293,183 @@ def unpack_annotation(
...
+                if self.max:
+                    if option.max_value is None:
+                        option.max_value = self.max
+
+        if isinstance(value, tuple):
+            if len(value) != 2:
+                raise ValueError("Range can only take 1-2 arguments")
+
+            min = value[0]
+            max = value[1]
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3376]: These names shadow built-ins, so probably best to rename

File: nextcord/application_command.py [hunk 8]
```diff
@@ -3276,3 +3293,183 @@ def unpack_annotation(
...
+            async def convert(self, interaction: Interaction, value: Any) -> Any:
+                return value
+
+            def modify(self, option: SlashCommandOption):
+                if self.min:
+                    if option.min_value is None:
+                        option.min_value = self.min
+                if self.max:
+                    if option.max_value is None:
+                        option.max_value = self.max
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3369]: Can do with less nesting
```suggestion
                if self.min and option.min_value is None:
                    option.min_value = self.min
                if self.max and option.max_value is None:
                    option.max_value = self.max
```

File: nextcord/application_command.py [hunk 9]
```diff
@@ -3276,3 +3293,183 @@ def unpack_annotation(
...
+            async def convert(self, interaction: Interaction, value: Any) -> Any:
+                return value
+
+            def modify(self, option: SlashCommandOption):
+                if self.min:
+                    if option.min_length is None:
+                        option.min_length = self.min
+                if self.max:
+                    if option.max_length is None:
+                        option.max_length = self.max
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3446]:
```suggestion
                if self.min and option.min_length is None:
                    option.min_length = self.min
                if self.max and option.max_length is None:
                    option.max_length = self.max
```

File: nextcord/application_command.py [hunk 10]
```diff
@@ -3276,3 +3285,181 @@ def unpack_annotation(
...
+            Inner.min = None
+        elif isinstance(min, (int, float)):
+            Inner.min = min
+        else:
+            raise TypeError("Range min must be int or float.")
+
+        if isinstance(max, EllipsisType):
+            Inner.max = None
+        elif isinstance(max, (int, float)):
+            Inner.max = max
```
Comment (ooliver1) [inline nextcord/application_command.py:3383]:
```suggestion
            min_value = value[0]
            max_value = value[1]
        else:
            min_value = None
            max_value = value

        if min_value is None or isinstance(min_value, EllipsisType):
            Inner.min = None
        elif isinstance(min_value, (int, float)):
            Inner.min = min_value
        else:
            raise TypeError("Range min must be int or float.")

        if isinstance(max_value, EllipsisType):
            Inner.max = None
        elif isinstance(max_value, (int, float)):
            Inner.max = max_value
```

File: nextcord/application_command.py [hunk 11]
```diff
@@ -3276,3 +3285,179 @@ def unpack_annotation(
...
+
+        if isinstance(value, tuple):
+            if len(value) != 2:
+                raise ValueError("String can only take 1-2 arguments")
+
+            min = value[0]
+            max = value[1]
+        else:
+            min = None
+            max = value
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3444]: Same thing as above about built-ins

File: nextcord/application_command.py [hunk 12]
```diff
@@ -3276,3 +3285,179 @@ def unpack_annotation(
...
+
+        if isinstance(min, EllipsisType):
+            Inner.min = None
```
Comment (DenverCoder1) [inline nextcord/application_command.py:3447]: Does this support min being None (if min is left out?) Seems different from Range above.

**Review by DenverCoder1 — APPROVED** [review]
Looks good to me. Everything tested and working now. :+1:

Only concern is with SlashOption not overriding the annotation when setting it to None, eg. `option: Range[1.1, 2.2] = SlashOption(min_value=None)`. I don't think this is a huge concern though, as it would be better, of course, for the developer to not write inconsistent code.

**Pull request merged**